0726-dom-element-descriptor-interface.md

```diff
@@ -0,0 +1,265 @@
+) {
+  if (!target) {
+    throw new Error('Must pass an element or selector to `click`.');
+  }
+
+  let element: Element | null;
+  if (target instanceof Element) {
+    element = target;
+  } else if (typeof target === 'string') {
+    let selector = target as string;
+    element = document.querySelector(selector);
+    if (!element) {
+      throw new Error(`element not found when calling \`click(${selector})\`.`);
+    }
+  } else {
+    // IElementDescriptor or IElementDescriptorProvider
+    let descriptor = (target as IElementDescriptorProvider)[ELEMENT_DESCRIPTOR];
+    if (!descriptor) {
+      descriptor = target as IElementDescriptor;
+    }
+
+    let element = descriptor.element;
+    if (!element) {
+      let description = descriptor.description;
+      if (description) {
+        throw new Error(
+          `element not found when calling \`click()\` with descriptor: ${description}`
+        );
+      } else {
+        throw new Error(
+          `element not found when calling \`click()\` with descriptor`
+        );
+      }
+    }
+  }
+
+  element!.click();
+}
+```
+
+To further illustrate, this is an (unnecessary) implementation of a descriptor and descriptor factory that just wrap a selector:
+
+```ts
+class SelectorDescriptor implements IElementDescriptor {
+  constructor(private selector: string) {}
+
+  get element() {
+    return document.querySelector(this.selector);
+  }
+
+  get elements() {
+    return document.querySelectorAll(this.selector);
+  }
+
+  get description(): string {
+    return this.selector;
+  }
+}
+
+class SelectorDescriptorProvider implements IElementDescriptorProvider {
+  public [ELEMENT_DESCRIPTOR]: IElementDescriptor;
+
+  constructor(selector: string) {
+    this[ELEMENT_DESCRIPTOR] = new SelectorDescriptor(selector);
+  }
+}
+```
+
+This RFC proposes to extend `@ember/test-helpers`'s DOM helpers and `qunit-dom`'s DOM assertions to accept arguments of type `IElementDescriptor` and `IElementDescriptorProvider` anywhere they accept a selector or an `Element`.
+
+### Why a symbol? Why two interfaces?
+
+It would be possible to only have the `IElementDescriptor` interface and not the `IElementDescriptorProvider` interface, and do away with the `ELEMENT_DESCRIPTOR` symbol. The short answer for why we have them is that one of this RFC's target use cases is page objects providing element descriptors, and if they _were_ element descriptors then the element descriptor interface properties/methods would be in the same namespace as user-defined properties on the page objects (e.g. a user might want to implement a `description` property on an album page object containing the text of the album's description, and this would conflict with the `description` property on the `IElementDescriptor` interface). So the `IElementDescriptorProvider` interface and `ELEMENT_DESCRIPTOR` symbol exist to allow such page objects to avoid namespace conflicts.
```

Comment:
Based on some discussion with rwjblue, I've modified the RFC to eliminate the symbol usage, and instead implement a registration/lookup API that stores the actual data in an unspecified manner (but we all know it would be a `WeakMap` 😉 ).

I believe this addresses your concern about the additional complexity of implement DOM helpers like `click()` -- they would only need to support a single interface for their arguments (in addition to the existing supported types), and I think it's pretty likely that the library described in this RFC would also provide one or more helper function to take care of the boilerplate of handling arguments of type `string | Element | null | IDOMElementDescriptor`.

Let me know what you think (and sorry for making such a big update to the RFC _right_ after you finished reading it 🤦 )

Reply:
Yes, definitely! `createDOMDescriptor` seems to be a good step forward to me. However, I'm still unsure about an ability to have plain classes as definitions. It doesn't seem to produce any new value if we have `createDOMDescriptor`.